### PR TITLE
OCPBUGS-1731: Merge 1.10.0

### PR DESCRIPTION
**Pull request by gcs278**
Rebased to CoreDNS 1.10.0 so that we use the Kubernetes 1.25 packages.

I made this PR using the same process as https://github.com/openshift/coredns/pull/73, https://github.com/openshift/coredns/pull/52, https://github.com/openshift/coredns/pull/60, and https://github.com/openshift/coredns/pull/70:

1. `git remote update`
2. `git checkout v1.10.0`
3. `git checkout -b merge-1.10.0-tmp`
4. `git checkout origin/master`
5. `echo 'merge coredns/coredns v1.10.0' | git commit-tree 'merge-1.10.0-tmp^{tree}' -p HEAD -p merge-1.10.0-tmp -F -`
     `ead3b6507d7bd2fc3e67b79d8ff0dff296396bac`
7. `git branch merge-1.10.0 ead3b6507d7bd2fc3e67b79d8ff0dff296396bac`
8. `git checkout merge-1.10.0`
9. `git rev-list --reverse --no-merges 555dd549ee36ba7f626cf8c53707c74af318ec09..origin/master | git cherry-pick --stdin`
    **Note**: `555dd549ee36ba7f626cf8c53707c74af318ec09` is the merge commit from the last merge https://github.com/openshift/coredns/pull/73
10. My merge commit from the above process was commit https://github.com/openshift/coredns/commit/ead3b6507d7bd2fc3e67b79d8ff0dff296396bac.

Before the rebase, we had the following commits:

```
% git --no-pager log --oneline --no-merges 555dd549ee36ba7f626cf8c53707c74af318ec09..origin/master
ad24eacf6 Updating coredns images to be consistent with ART Reconciling with https://github.com/openshift/ocp-build-data/tree/2e5eba0108299087402c119ba0dc979a0a6c0bd1/images/coredns.yml
349db7a33 UPSTREAM: <carry>: openshift: Revert "Revert "add wildcard warnings (#5030)" (#5167)"
461cbbf60 UPSTREAM: <carry>: openshift: Revert "remove wildcard query functionality (#5019)"
7a41d2b87 UPSTREAM: <carry>: openshift: Disable dependabot
1d2d4f767 UPSTREAM: <carry>: openshift: Vendor required modules
3dc197d15 UPSTREAM: <carry>: openshift: .gitignore: Properly ignore coredns binary
9176cd5bf UPSTREAM: <carry>: openshift: Make coredns exclusion more specific
488cc73b1 UPSTREAM: <carry>: openshift: Add a `make test` target
b648be195 UPSTREAM: <carry>: openshift: Add product build config
1e6995352 UPSTREAM: <carry>: openshift: Add OWNERS
% 
```

I edited the "Vendor required modules" commit by running go mod vendor and go mod tidy and ammend commit `1d2d4f767 UPSTREAM: <carry>: openshift: Vendor required modules`.

I squashed/merged the following commit into the existing "Add product build config" commit:
```
ad24eacf6 Updating coredns images to be consistent with ART Reconciling with https://github.com/openshift/ocp-build-data/tree/2e5eba0108299087402c119ba0dc979a0a6c0bd1/images/coredns.ym
```

Git merge conflicts for `461cbbf60 UPSTREAM: <carry>: openshift: Revert "remove wildcard query functionality (#5019)"`:
**plugin/kubernetes/kubernetes.go**
  - Line 630: `matchPortAndProtocol` function comment update, but `matchPortAndProtocol` function has been removed so I ignored the updates (see [PR Delta](https://github.com/coredns/coredns/pull/5592/files#diff-afedb9efe6b2bbd5feec11592f7b3fed72d6355525ad50be0d5812e154b6f1eaR596))

**plugin/kubernetes/external.go**
  - Line 52: changes to variables declaration conflicted because they added `endpoint` as a string. I merged by taking upstream variable declarations, but added `port = "*"` and `protocol = "*"` assignment below for reversion (see [PR Delta](https://github.com/coredns/coredns/pull/5505/files#diff-adb4b352bbef770140afca16f60005e017616dad9fa090f773ba93238d6e1863R52))
  - Line 114 & 134: Refactoring moved calls to `matchPortAndProtocol` and added it in an additional location for resolving headless services (see [PR delta](https://github.com/coredns/coredns/pull/5505/files#diff-adb4b352bbef770140afca16f60005e017616dad9fa090f773ba93238d6e1863R109))
     - I'm assuming we want wildcard support for headless and that wildcard support is remove via `matchPortAndProtocol`, but we don't have that function anyways so just using the `match` function.

### CODE REVIEW

**Comment by gcs278** [comment]
/hold
need to merge a carry patch

File: plugin/kubernetes/external.go [hunk 1]
```diff
@@ -11,24 +11,50 @@ import (
...
-	protocol := "*"
+	// We are dealing with a fairly normal domain name here, but we still need to have the service,
+        // namespace and if present, endpoint:
+        // service.namespace.<base> or
+        // endpoint.service.namespace.<base>
+        var port, protocol, endpoint string
+        //
+        // for service (and SRV) you can also say _tcp, and port (i.e. _http), we need those be picked
+        // up, unless they are not specified, then we use an internal wildcard.
+        port = "*"
```
Comment (gcs278) [inline plugin/kubernetes/external.go:56]: this was a tricky merge

File: plugin/kubernetes/external.go [hunk 2]
```diff
@@ -67,51 +101,141 @@ func (k *Kubernetes) External(state request.Request) ([]msg.Service, int) {
...
 					continue
 				}
-				rcode = dns.RcodeSuccess
-				s := msg.Service{Host: ip, Port: int(p.Port), TTL: k.ttl}
-				s.Key = strings.Join([]string{zonePath, svc.Namespace, svc.Name}, "/")
 
-				services = append(services, s)
+				for _, eps := range ep.Subsets {
+					for _, addr := range eps.Addresses {
+						if endpoint != "" && !match(endpoint, endpointHostname(addr, k.endpointNameMode)) {
```
Comment (gcs278) [inline plugin/kubernetes/external.go:116]: Another tricky merge

File: plugin/kubernetes/external.go [hunk 3]
```diff
@@ -67,51 +101,141 @@ func (k *Kubernetes) External(state request.Request) ([]msg.Service, int) {
...
+							continue
+						}
+
+						for _, p := range eps.Ports {
+							if !(match(port, p.Name) && match(protocol, string(p.Protocol))) {
```
Comment (gcs278) [inline plugin/kubernetes/external.go:121]: Another tricky merge

**Comment by gcs278** [comment]
/unhold
I think I resolved the merges

**Comment by cybertron** [comment]
I tried this out locally to verify all the on-prem stuff still works with it and everything looks good.

**Comment by gcs278** [comment]
ci/prow/e2e-gcp-serial: OVNKubernetesControllerDisconnectedSouthboundDatabase Issue, disruption and alert issues.
ci/prow/e2e-metal-ipi: Lots of failures everywhere, something went wrong maybe with the install

**Comment by gcs278** [comment]
e2e-gcp-serial: `ingress-to-oauth-server-new-connections was unreachable during disruption testing for at least 54m3s of 2h4m53s (maxAllowed=9s):` type stuff
e2e-metal-ipi: again, lots of failures everywhere, something went wrong maybe with the install
/retest

**Comment by gcs278** [comment]
FYI I have never seen a successful e2e-metal-ipi for this repo. But it's option.
ci/prow/e2e-gcp-serial: Both OVN Database and Disruption issues...again. Distruption issues are supposed to be fixed, but not seeing that yet.
/retest

**Comment by cybertron** [comment]
I'm not sure the metal-ipi job definition is up to date on this repo. I tested this locally in a metal ipi deployment and it worked, so I don't think we need to worry about that.

I'll write a patch to update the job definition on this repo to match the other places metal-ipi runs.

**Comment by gcs278** [comment]
@cybertron Thanks for the fix for metal-ipi. I think we are aiming to get this in soon (maybe tomorrow or early next week) so if it's fixed by then, we can try to get a clean CI run, if not, I'll take your word on "we don't need to worry about it".

**Review by brandisher — COMMENTED** [review]
Overall the changes look good and match what we discussed yesterday. I left one question for @Miciah about whether we need a release note for the change in SERVFAIL response caching; I'm assuming that SERVFAILs were not cached previously and will now be cached for 5s.

**Comment by gcs278** [comment]
Just OVN alerts now.

**Comment by gcs278** [comment]
/retest ci/prow/e2e-metal-ipi
May be fixed now.

**Comment by gcs278** [comment]
/test ci/prow/e2e-metal-ipi
may be fixed now.

**Comment by gcs278** [comment]
/test e2e-metal-ipi
may be fixed now.

**Comment by cybertron** [comment]
/test e2e-metal-ipi

I'm seeing a lot of io timeouts in the coredns logs and it looks like the whole deployment timed out. Maybe the test node was having issues?

**Comment by brandisher** [comment]
`OVNKubernetesControllerDisconnectedSouthboundDatabase`
/test e2e-gcp-serial

**Comment by gcs278** [comment]
e2e-metal-ipi worked :heavy_check_mark: thanks @cybertron 

e2e-gcp-serial is a lost cause given the outstanding OVN and disruption bugs. I think we need an override now.

File: plugin/cache/cache.go [hunk 4]
```diff
@@ -59,6 +65,7 @@ func New() *Cache {
 		ncache:     cache.New(defaultCap),
 		nttl:       maxNTTL,
 		minnttl:    minNTTL,
+		failttl:    minNTTL,
```
[the remark this answers is not in the document]
Comment (Miciah) [reply plugin/cache/cache.go:68]: SERVFAIL was cached for 5s in CoreDNS 1.9.2.
https://github.com/coredns/coredns/blob/092c144491e7f62d71eda98a07e4ce815d9df035/plugin/cache/cache.go#L159-L161
This change only makes it configurable, but the default is 5 seconds, and we aren't changing the default.

File: plugin/kubernetes/external.go [hunk 5]
```diff
@@ -46,10 +46,15 @@ func (k *Kubernetes) External(state request.Request, headless bool) ([]msg.Servi
...
-	var port, protocol, endpoint string
+        // namespace and if present, endpoint:
+        // service.namespace.<base> or
+        // endpoint.service.namespace.<base>
+        var port, protocol, endpoint string
+        //
+        // for service (and SRV) you can also say _tcp, and port (i.e. _http), we need those be picked
+        // up, unless they are not specified, then we use an internal wildcard.
+        port = "*"
+        protocol = "*"
```
[the remark this answers is not in the document]
Comment (gcs278) [reply plugin/kubernetes/external.go:57]: I think you mean spaces vs. indents? Great catch, I didn't see that my edits used spaces in github diff or in my editors.
fixed

File: plugin/cache/cache.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (brandisher) [reply plugin/cache/cache.go:68]: Ah okay, thanks for catching that and the link! That's what I get for making an assumption :-)

**Comment by Miciah** [comment]
In the course of reviewing the wildcard carry patch (where @gcs278 had to resolve some conflicts for this rebase), I found and reported https://github.com/coredns/coredns/pull/5657.  If I apply https://github.com/coredns/coredns/pull/5657 on top of https://github.com/openshift/coredns/pull/76, then this test case for `TestExternal` fails:
```go
	{
		Qname: "*._not-udp-or-tcp.svc1.testns.example.com.", Rcode: dns.RcodeSuccess,
	},
```
However, the same test fails if I check out the v1.8.7 tag (the last upstream version that supported wildcard lookups) and fix `TestExternal` there.  I believe that this test case was always incorrect:  The test case has a query for a wildcard with no match, so it should expect an error response.  

Additionally, I tried the following two test cases:
```go
	{
		Qname: "any._tcp.svc1.testns.example.com.", Rcode: dns.RcodeSuccess,
		Msg: []msg.Service{
			{Host: "1.2.3.4", Port: 80, TTL: 5, Key: "/c/org/example/testns/svc1"},
		},
	},
	{
		Qname: "_any._tcp.svc1.testns.example.com.", Rcode: dns.RcodeSuccess,
		Msg: []msg.Service{
			{Host: "1.2.3.4", Port: 80, TTL: 5, Key: "/c/org/example/testns/svc1"},
		},
	},
```
These test cases both pass on v1.8.7 with `TestExternal` fixed and on https://github.com/coredns/coredns/pull/5657, so I believe there has always been a bug in the wildcard logic (that is, both "any" and "_any" are treated as wildcards), and this PR maintains bug-for-bug compatibility.

Let's get this PR merged and let it soak.  

If upstream accepts https://github.com/coredns/coredns/pull/5657 and someone asks for us to fix the bug in openshift/coredns, then we can treat it as a bug like any other and fix it with a carry patch for https://github.com/coredns/coredns/pull/5657.  

/lgtm
/approve

**Comment by Miciah** [comment]
[e2e-aws-upgrade failed](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_coredns/76/pull-ci-openshift-coredns-master-e2e-aws-upgrade/1575902353813409792) with the following failure:
```
[sig-arch] events should not repeat pathologically
{  1 events happened too frequently  event happened 27 times, something is wrong: ns/openshift-console-operator pod/console-operator-55ff985f74-zbf64 node/ip-10-0-155-208.ec2.internal - reason/ProbeError Readiness probe error: Get "https://10.130.0.35:8443/readyz": dial tcp 10.130.0.35:8443: connect: connection refused body:  }
```
The closest match for this failure that I found is [BZ#1990419 [sig-arch] events should not repeat pathologically - event happened more than 20 times, something is wrong: ns/openshift-console pod/console-XXX node/XXX - reason/Unhealthy Readiness probe failed](https://bugzilla.redhat.com/show_bug.cgi?id=1990419), which was closed due to insufficient data.  Let's see if the same failure re-occurs.  

/test e2e-aws-upgrade

**Pull request merged**